File: src/Shared/Table.js
import React from 'react';
import './Table.css';
import '../common.css';
import {Route, NavLink, HashRouter, BrowserRouter, Link} from "react-router-dom"

export default class Table extends React.Component {
    constructor(props)
    {
        super(props);
    }
    
    componentDidMount() {
    }
    componentWillUnmount() {
    }

    rowClicked(data) {

    }
    render () {
        return (
            <table>
                <thead class={"header"}>
                    <tr>
                        {Object.keys(this.props.columns).map((column, i) => {
                            return (
                                <th className={"header-cell header-cell-alignLeft " + (i == 0 ? "primary-cell" : "")}>
                                    <span>
                                        {column}
                                        <ArrowIcon/>
                                    </span>
                                </th>
                            )
                        })}
                    </tr>
                </thead>
                <tbody>
                <BrowserRouter basename={this.props.baseUrl}>
                    {this.props.data.map((data, i) => {
                        return (
                                <tr key={i} className={"body-row"} >
                                    {
                                        Object.values(this.props.columns).map((column, j) => {
                                            return (
                                                <td class={j == 0 ? "primary-cell" : ""}>
                                                    <Link to={`/${i}`}>
                                                        <div>{data[column]}</div>
                                                    </Link>
                                                </td>
                                            )
                                        })
                                    }
                                </tr>
                            )
                    })
                    }
                </BrowserRouter>

                </tbody>
            </table>
        )
    }
}

function ArrowIcon(props)
{
    function getDirectionClass()
    {
        switch(props.direction)
        {
            case "descending":
                return "arrow-icon-descending"
            case "ascending":
                return "arrow-icon-ascending"
            default:
                return ""
        }
    }

    return (
        <svg className={"arrow-icon " + getDirectionClass()} focusable="false"
             viewBox="0 0 24 24" aria-hidden="true" role="presentation">
            <path d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"/>
        </svg>
    )
}

File: ReactK/src/main/kotlin/Table.kt
import kotlinx.html.classes
import react.*
import react.dom.*

external interface TableProps : RProps {
    var columns : List<String>
    var data : List<List<String>>
}

class Table : RComponent<TableProps, RState>(){
    override fun RBuilder.render() {
        table {
            thead {
                tr {
                   for (column in props.columns)
                    {
                        th {
                            val index = props.columns.indexOf(column)
                            key = index.toString()

                            attrs["className"] = "header-cell header-cell-alignLeft"
                            if (index == 0)
                            {
                                attrs["className"] += " primary-cell"
                            }
                            span {
                                +column
                            }
                        }
                    }
                }
            }
            tbody {
                props.data.forEachIndexed{ i, it ->
                    tr {
                        attrs["className"] = "body-row"
                        key = i.toString()
                        it.forEachIndexed { j, it ->
                            td {
                                key = j.toString()
                                if (j == 0)
                                    attrs["className"] = "primary-cell"
                                link {
                                    +it
                                }
                            }
                        }
                    }
                }
            }
        }
    }
}
external interface SortArrowProps : RProps {
    var direction : String?
}
class ArrowIcon : RComponent<SortArrowProps, RState>() {
    fun getDirectionClass() : String {
        return when(props.direction)
        {
            "descending" -> "arrow-icon-descending"
            "ascending"-> "arrow-icon-ascending"
            else -> ""
        }
    }
    override fun RBuilder.render() {
        svg (classes = "arrow-icon ${getDirectionClass()}"){

        }
    }
}